Add In HA Event Callback

byterock opened the pull request:
Had this patch lying about on an old box.

here is a link on the ORacel site

[https://docs.oracle.com/database/121/ADFNS/adfns_avail.htm#ADFNS920](url)

and the OCI docs

[https://docs.oracle.com/cd/E11882_01/appdev.112/e10646/oci09adv.htm#LNOCI16691](url)

Will add it in later this wee

daglem committed:
Fix warnings from gcc/clang, correct a few typos

daglem committed:
Avoid use of inactive database handle in statement destructor

This error caused a SIGSEGV in t/51scroll.t with ORA_DRCP=1

daglem committed:
Plug various environment handle leaks

This fixes https://github.com/pythian/DBD-Oracle/issues/27 - among other cases. Note that with ithreads, the global imp_drh environment handle still leaks.

The previous error paths and the attempted partial fix in https://github.com/pythian/DBD-Oracle/issues/28 could cause crashes because of missing curly brackets in if-statements calling OCIHandleFree_log_stat (which is a macro).

The addition of a driver handle destructor also prepares for timely destruction of DRCP session pools. This requires DBI >= 1.623, which adds logic to force destruction of children before parents during global destruction.

daglem committed:
Correct setting of default driver name

Don't set OCI_ATTR_DRIVER_NAME on an unallocated session handle, potentially causing corruption or crashes.

Set the default driver name to "DBD::Oracle : <version>", using the de facto standard driver naming scheme.

This fixes https://github.com/pythian/DBD-Oracle/issues/32

daglem committed:
Add configuration for Oracle Fast Application Notification (FAN)

Facilitate setting of OCI_EVENTS in a new OCI environment via the attribute ora_events or the environment variable ORA_EVENTS.

daglem committed:
Add configuration for Oracle Fast Application Notification (FAN)

Facilitate setting of OCI_EVENTS in a new OCI environment via the attribute ora_events or the environment variable ORA_EVENTS.

daglem committed:
Rewrite DRCP session pooling to work as intended

Rewrite of DRCP session pooling, influenced by PHP's oci8.c

Enable sharing (reuse) of server sessions, and implement caching of the client side of the session pool(s) between connects, keeping the connection(s) to the DRCP Connection Broker. Depending on server hardware and network latency, subsequent connects are now up to several hundred times faster than before the rewrite.

Use session tagging to fetch shared server sessions with correct NLS character sets.

The attribute ora_drcp_rlb or the environment variable ORA_DRCP_RLB controls runtime connection load balancing for Oracle RAC.

Documentation update.

All tests pass with ORA_DRCP=1, with the exception of t/60reauth.t (it goes without saying that reauthentication does not work with session pooling).

(updated ocitrace.h - Dean Hamstead)

daglem committed:
Get correct "tag found" status from OCISessionGet

The caller must pass non-NULL pointers for "retTagInfo" and "retTagInfo_len".

deanpearce committed:
Merge pull request #50 from whindsx/master

Have ora_libdir check $ORACLE_HOME/lib/64 when it exists

deanpearce committed:
Merge pull request #51 from pythian/feature/ora-version

Fix for Oracle version default value.

Commit:
Fix awkwardness and spelling

moregan committed:
Change misspelling of SH_LIBRARY_PATH in quoted text

moregan committed:
stop referring to the VARCHAR2 as deprecated

Commit:
Fix for GH36

djzort commented:
Has this been merged? Its confusing whats happened?

Commit:
Fix in/from #56

byterock commented:
Would of been nice if it ever was 

Unfortunerly  I no longer have access to the Box with the set up to test the orginal code I sent Dean and Yanick the Code way back in 2012 Not sure if it ever got in.

I could try and recreate it as I have the patch about just that I have no access to an Oracle system with HA on more that 1 box to test it against the current code base

djzort commented:
Just in the changelog of this PR its not clear what your diff is. If you have the .patch file, perhaps just post it here?

Commit:
Fix up for GH36

Commit:
Fix compile warnings about int vs long unsigned

Commit:
https://lintian.debian.org/tags/spelling-error-in-manpage.html etc

Commit:
Make a passing reference to other authors

Commit:
Tidy up TODO

Commit:
Rename from Todo to TODO

djzort committed:
Overhaul tests, dzil usage, perltidy and tweaks

Commit:
Update mailmap file

Commit:
Add Travis CI file

Commit:
Update changes based on included PR's and GH bugs

Commit:
Fix #57 based upon suggestions from @tonycoz.

See https://github.com/pythian/DBD-Oracle/issues/57#issuecomment-414253275

djzort committed:
Update changelog for GH#57

djzort committed:
Update Changes file for 1.75_42

whindsx committed:
Find ORACLE_HOME based off Operating System (Linux) when other attempts in
find_oracle_home() have failed

Determine Oracle client version from ORACLE_HOME path when other attempts
in get_client_version() have failed

Commit:
Bump to 1.75

Commit:
Adjust NextRelease to be consistent with Changes format we're using

Commit:
Bump to v1.76

mjegh committed:
Merge pull request #62 from ByteFoundryAU/winc

Various PR consolidated

tonycoz committed:
correct the type passed to OCIAttrGet() for OCI_ATTR_TAF_ENABLED

should fix #35

mjegh committed:
Merge pull request #78 from tonycoz/fix35

correct the type passed to OCIAttrGet() for OCI_ATTR_TAF_ENABLED

CarstenGrohmann committed:
A few typos

djzort committed:
Remove Authority section as PYTHIAN is no longer involved

mjegh committed:
Merge pull request #83 from ByteFoundryAU/update-authority

Remove Authority section as PYTHIAN is no longer involved

mjegh committed:
Merge pull request #81 from CarstenGrohmann/Typos

A few typos

djzort committed:
Minor adjustments to README file

djzort committed:
Keep all lines under 80 columns

mjegh committed:
Merge pull request #88 from ByteFoundryAU/updatereadme

Minor adjustments to README file

CarstenGrohmann committed:
Move freeing global envhp to destruction of last dbh

As the order of object destruction during "global destruction" is
undefined. DBI traces show that the global environment is freed in the
driver destruction process before all database connections are closed
and their resources freed. This causes the segfault in #65.

This change moved freeing of the global environment handle from the
destruction of the driver handle (dbd_dr_destroy()) to the destruction
of the database handle (dbd_db_destroy()).

This implementation closes the global environment handle during the DBI
shutdown process if the last database handle uses the global
environment handle.

If the last database handles uses a session environment handle, the
global environment handle will not be destroyed.

djzort committed:
Enable testing against 5.28

djzort committed:
Remove disconnect() which should no longer be needed

CarstenGrohmann committed:
A few typos more

mjegh committed:
Merge pull request #91 from CarstenGrohmann/More_typos

A few typos more

mjegh committed:
Merge pull request #74 from whindsx/make-linux-rpm

Improve finding ORACLE_HOME when installing on Linux OS

CarstenGrohmann committed:
Enable travis helpers DEBUG options

djzort committed:
Try the dzil build command again manually

djzort committed:
Try something different

djzort committed:
Try excluding cpanfile stuff

djzort committed:
Revert "Try excluding cpanfile stuff"

This reverts commit a83998a5f95d40f49cdce28afe51773fc49cd09f.

djzort committed:
Try excluding Git::Check

Commit:
Avoid GCC-ism so that HP-UX can compile. GH#92

Commit:
Revert "Try excluding Git::Check"

This reverts commit ae237d5a3f129824a0ff5ec8ed0976283e6e5b10.

Commit:
Try disabling Git::CommitBuild as recommended by @kentnl

mjegh committed:
Merge remote-tracking branch 'refs/remotes/perl5-dbi/master'

mjegh committed:
disable Git::CommitBuild

mjegh committed:
Merge pull request #96 from mjegh/master

Master

mjegh committed:
Merge pull request #95 from ByteFoundryAU/gh92

Avoid GCC-ism so that HP-UX can compile. GH#92

mjegh committed:
Delete dist.ini

Already applied

mjegh committed:
remove dist.ini change

mjegh committed:
Merge branch 'ByteFoundryAU-destroy_envhp_with_last_dbh'

mjegh committed:
Merge remote-tracking branch 'refs/remotes/perl5-dbi/master'

mjegh committed:
Merge pull request #97 from mjegh/master

Destroy envhp with last dbh

mjegh committed:
Merge pull request #90 from ByteFoundryAU/testp28

Enable testing against 5.28

djzort committed:
Remove debug line, this wasnt supposed to go to master

djzort committed:
Remove Git things we arent using or worried about

mjegh committed:
Merge pull request #98 from ByteFoundryAU/p5master

Clean ups on dist.ini and .travis.yml

djzort committed:
Update github path

djzort committed:
Add another mailmap entry for Dean Hamstead

djzort committed:
Update mailmap file

mjegh committed:
up version for deve release

mjegh committed:
Merge remote-tracking branch 'refs/remotes/perl5-dbi/master'

mjegh committed:
Merge pull request #100 from mjegh/master

up version

mjegh committed:
Merge pull request #99 from ByteFoundryAU/p5master

Update github path in CONTRIBUTING.mkd

djzort committed:
Remove x flag

djzort committed:
Remove references to long gone EXAMPLE FILES section

djzort committed:
Unlink 'LOB Locator Method Interface' as there is no section for it

mjegh committed:
Merge pull request #102 from ByteFoundryAU/kwalitee_1_77_1

Kwalitee 1.77_1

mjegh committed:
Bump to 1.78

mjegh committed:
Merge remote-tracking branch 'refs/remotes/perl5-dbi/master'

mjegh committed:
Merge pull request #104 from mjegh/master

Bump to 1.78

mjegh committed:
Bump version

mjegh committed:
Bump to 1.791

mjegh committed:
Run orphaned test code

mjegh committed:
Updated bugtracker to git
Updated home page to metacpan
Updated Changes
Updated dis.ini to bump to 1.792

djzort committed:
Adjust prerequisites in dist.ini

mjegh committed:
Update version to 1.80 because it causes packagers a problem

CarstenGrohmann committed:
Extend documentation about leaking handles

and rename leak_state to leak_handles for better understanding.

alexm committed:
Build DBD::Oracle on Debian systems without LD_LIBRARY_PATH

Patch based on a proposed fix by Slaven Rezić <eserte>
in https://github.com/perl5-dbi/DBD-Oracle/issues/72
and improved by me afterwards.

The patch has been successfully applied to Debian for a
few releases now:

https://salsa.debian.org/perl-team/modules/packages/libdbd-oracle-perl/blob/master/debian/patches/build-without-ld-libray-path.patch

alexm committed:
Apply fix to debian only

As suggested by Wesley Hinds <whindsx>:
https://github.com/perl5-dbi/DBD-Oracle/pull/112#issuecomment-559244350

djzort committed:
Add $OH/include as a possible header location

djzort committed:
Add perl 5.30 to travis

djzort committed:
Remove dead links from README

djzort committed:
Adjust deprecated .travis.yml syntax

djzort committed:
Add perl 5.32

djzort committed:
Tweaks to dist.ini to make the release cleaner

djzort committed:
Remove POD sections that are now being inserted automatically

djzort committed:
Small adjustments to the script that installs oracle xe for testing

djzort committed:
spit out the sha256 of oracle-xe. maybe check it later

djzort committed:
Add a Dockfile to help people hack on DBD::Oracle quickly

djzort committed:
Update comments in Dockerfile

djzort committed:
Add some debugging

djzort committed:
Move t/README to /TESTING.md for GH#119

djzort committed:
Update README with some minor improvements

djzort committed:
Update markdown in TESTING.md

djzort committed:
Dont test 5.32 as Travi's perl install for it seems broken

djzort committed:
Mention mkta.pl in TESTING.md

djzort committed:
Update project links in Makefile.PL

djzort committed:
Update MANIFEST

djzort committed:
Add instantclient 18.3 and 18.5 to travis, use xenial instead of trusty

djzort committed:
Add instantclient 19.6 to travis

djzort committed:
Ensure strict and warnings are first

djzort committed:
Swap tabs to spaces in Makefile.PL

whindsx committed:
Run nonassigned pl/sql tests based off versions known to have the issue.
GH#70.

alexm committed:
Hint ld that libclntsh deps are needed

Otherwise the references to the extra libs needed by -lclntsh would be
removed and LD_LIBRARY_PATH would be necessary again.

Tux committed:
Oracle Instant Client 21 has been released

hakonhagland committed:
Add rpath to ORACLE_HOME to DBD/Oracle/Oracle.bundle

Fixes issue #128 by adding rpath to ORACLE_HOME to DBD/Oracle/Oracle.bundle.

Commit:
fix materialised views being misclassified as tables in ->table_info - github #132

Commit:
Fix bugtracker ref in dist metadata github #130

djzort committed:
Update Changes

djzort committed:
v1.81

  [BUG FIXES]

  Run nonassigned pl/sql tests based off versions known to have the issue. - (GH#70, Wesley Hinds)
  Oracle Instant Client 21 support - (Tux)
  Add rpath to ORACLE_HOME to DBD/Oracle/Oracle.bundle  - (GH#128, hakonhagland)
  fix materialised views being misclassified as tables in ->table_info - (GH#132, John Smith)
  Fix bugtracker ref in dist metadata github - (GH#130, John Smith)

  [MISCELLANEOUS]

  Updates to POD and README file
  Moved old README file to t/TESTING.md
  Updated links

Commit:
fix github metadata woes as best as possible

For some reason the github dzil plugin is not fining the issues page, so
for now adding it to Meta.yml manually - i.e. `bugtracker:
https://github.com/perl5-dbi/DBD-Oracle/issues` on release will have to do.

djzort committed:
The CPAN doesnt recognize .mkd as markdown. Move to .md

djzort committed:
Note merged PR

Commit:
prep 1.83 for release

whindsx committed:
Updated list of OCI versions known to have bind_param_inout
nonassigned bug. These versions will be skipped in 25plsql.t.

Originally reported in GH#70.

whindsx committed:
Include 21.4 in OCI versions known to have bind_param_inout
nonassigned bug.

Originally reported in GH#70.

whindsx committed:
Check for appropriate permissions before running 28array_bind.t tests.

djzort committed:
Update changes for GH#140

djzort committed:
Update Changes for GH#141

whindsx committed:
Have find_headers() also search for header files based off the
major version number.

whindsx committed:
Check for appropriate permissions before running 56embbeded.t tests.

kjetillll committed:
Update 25plsql.t

Tested with instantclient version 21.5 and also this version should be added to @bad_oci_vers.

kjetillll committed:
Update Oracle.pm

djzort committed:
Update Changes from serveral merged PR

deanpearce committed:
Update .mailmap

Removing outdated information from the list.

djzort committed:
Update MANIFEST

djzort committed:
Update mailmap

avorop committed:
Rewrite of login6

This changes the way things are done during login6 (connect). In
particular:

1. OCIEnv is cached based on character sets and mode. One copy
is used in all threads (so OCI_THREADED is important!)

2. There are no more global variables for character sets. So each
connection uses exactly what is specified at connection time. Thus
option ora_envhp is removed.

3. Improved support for DRCP. Now the pools are independent from
threads. The pool is defined by DSN and UID. No more leaking of
connections or pools. When last imp_drh goes away the pools
are released. It is now also possible to tag sessions in addition
to using connection_class (the latter works only with DRCP).

4. Reauthenticate now keeps old session as long as possible, so
if new identity is wrong, then old session stays intact.

5. Support for using shared variables is also improved. There is
now additional function ora_shared_release that cleans up shared
connection, when it is not needed anymore. Though it is still
cleaner to used DRCP. No more handles are leaking.

6. Support for imp_take is improved. No handles are leaking.
But it is cleaner to use DRCP.

7. Small fixes for compilation and handling of attributes.

djzort committed:
Update changes for PR#150

djzort committed:
Bump version to dev of v1.90

djzort committed:
Swap from MetaYAML to MetaJSON

djzort committed:
Fix up bugtracker data in META information

djzort committed:
Bump docker image to 5.32.1

whindsx committed:
Move from TravisCI to Github Actions (GHA).
The GHA will build against the "latest" Ubuntu/Perl/OracleXE/InstantClient
that is currently available.

avorop committed:
Add protection for older OCI

This version uses to defines from oci.h, which where added after
version 12. Since I don't know exact version, when they were added,
I protect their usage by check for OCI above 18. For older OCI
these variables will be unavailable.

Additionally I've tried to fix some warnings in Makefile.PL and
stopped t/25plsql.t from checking if OCI has bug. After all DBD::Oracle
does not protect against this bug, so there is no sense to check if
OCI has it.

djzort committed:
Update changelog and bump version to 1.90_2

avorop committed:
Update dbdcnx.c

Added freeing of allocated handle. It is not needed anymore since OCIAttrSet copies its conent (according to Oracle developers)

djzort committed:
Update Changes, bump version

djzort committed:
Update changes for PR#152

avorop committed:
Convert inline functions to macros

Some old systems can not use inline function

djzort committed:
Add DBI to ConfigureRequires, BuildRequires, and Test Requires - (GH#155)

djzort committed:
Bump to v1.90_4

whindsx committed:
dbivport.h should be installed as well. GH#159

djzort committed:
Update changes file and bump version to 1.90_5

djzort committed:
Update README. Point to ::Troubleshooting:: and remove files long
deleted GH#156

haarg committed:
specify resources metadata explicitly in dist.ini

The latest stable release of DBD::Oracle listed the wrong bug tracker
and repository because it was released from a fork, and the [GithubMeta]
plugin used that fork to set the metadata.

Remove that plugin, and instead specify the bug tracker and repository
explicitly.

thesamesam committed:
Makefile.PL: help ora_libdir find ORACLE_HOME on Gentoo

On Gentoo, for better or worse, we install oracle-instantclient on 64-bit
systems to /usr/lib64/oracle/client/ with:
* /usr/lib64/oracle/client/lib for 32-bit libraries (multilib)
* /usr/lib64/oracle/client/lib64 for native 64-bit libraries

Adapt the logic to handle lib64 if it exists and we're using a 64-bit Perl.

(And fix some whitespace while at it.)

djzort committed:
It's now 2024

djzort committed:
Update changes for GH#162

djzort committed:
Bump version to 1.90

lzsiga committed:
Changing 'return' to block statement, fixing a problem created while transforming inline function to macro

djzort committed:
Remove oracle v21 specific code

whindsx committed:
Makefile.PL fix for missing Test::NoWarnings build requirement.
Along with appropriate update to Github Action.

Also we don't care about ExtUtils::MakeMaker version number anymore.

This fixes GH#175.

fedeliallalinea committed:
Makefile.PL: help to find linking libraries on Gentoo

On Gentoo we install oracle-instantclient libraries in
/usr/lib64/oracle/client/lib64 on 64-bit systems and in
/usr/lib/oracle/client/lib on 32-bit systems.

This was done because in a 64-bit system with multilib,
both 32-bit libraries (/usr/lib64/oracle/client/lib) and
64-bit libraries (/usr/lib64/oracle/client/lib64) were
installed. For now, this specification is disabled because
it is rare to have the same version of oracle instantclient
for both 32-bit and 64-bit.

djzort committed:
Update Changes file with recent commits

Tux committed:
Remove outdated #ifdefs

As the minimum required version of DBI is 1.623, these

when the versions stored in these variables change in DBI

ancientwizard committed:
- Minor housekeeping while investigating SEGV issue.
- SEGV's happen during the call to OCISessionEnd()
  (based on my usecase)
   o WHEN there are three or more threads
   o Connected to the same DB user at the same time
   o When the last thread makes the call to OCISessionEnd() the SEGV
     is created.
- I have discovered 05/05(today) that the SEGV can be iliminated (avoided)
   o WHEN the last thread to connect is the first to OCISessionEnd()
   o I'm going to write another smaller Perl script to reproduce
     the SEV and its avoidance.

ancientwizard committed:
Most of this is ancient news and clutter.
- messages have been released to seek a new life by going into the "LIGHT"

ancientwizard committed:
Added test specific to SEGV issue. t/90-segv.t
Noted that t/15threads.t randomly SEGV's

This happens a lot.
t/16cached.t .............. Can't call method "prepare" on an undefined value at .../t/cache2.pl line 21.

I figure others are seeing these same issues as I am.

t/rt74753-utf8-encoded.t .. 1/3 Issuing rollback() due to DESTROY without explicit disconnect() of DBD::Oracle::db handle ORA19C at t/rt74753-utf8-encoded.t line 86.
t/rt74753-utf8-encoded.t .. ok
t/rt85886.t ............... 1/2
SV = PV(0x5aa9e32f1c60) at 0x5aa9e349c358
  REFCNT = 1
  FLAGS = (POK,pPOK)
  PV = 0x5aa9e34b4590 "1"\0
  CUR = 1
  LEN = 16
t/rt85886.t ............... Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/2 subtests

Test Summary Report
-------------------
t/16cached.t            (Wstat: 256 (exited 1) Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
t/90-segv.t             (Wstat: 139 (Signal: SEGV, dumped core) Tests: 37 Failed: 0)
  Non-zero wait status: 139
  Parse errors: No plan found in TAP output
t/rt85886.t             (Wstat: 256 (exited 1) Tests: 2 Failed: 1)
  Failed test:  2
  Non-zero exit status: 1
Files=44, Tests=1492, 237 wallclock secs ( 0.22 usr  0.19 sys +  3.70 cusr  2.10 csys =  6.21 CPU)
Result: FAIL
Failed 3/44 test programs. 2/1492 subtests failed.
make: *** [Makefile:1114: test_dynamic] Error 1

ancientwizard committed:
I believe I found the primary bug which was also likely creating a memory leak.
I've left this better than I found it. But not as clean as I would like.
This isn't mine to own or make such decisions. But ...
 Parhaps it's time to abandon all that old Oracle Support.
(Some how I doubt it's been tested in a long time!)
Require perl-5.30.3+ & Oracle 17+. Clean up and simplify. Call it 2.0!
The existing threads test no longer SEGV's and the test I added also passes.

There are a couple of preexisting test fails I'll leave to the maintainers.

- God Speed!

t/rt85886.t ............... 1/2
SV = PV(0x6479e78dba50) at 0x6479e797ea60
  REFCNT = 1
  FLAGS = (POK,pPOK)
  PV = 0x6479e7a88350 "1"\0
  CUR = 1
  LEN = 16
t/rt85886.t ............... Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/2 subtests

Test Summary Report
-------------------
t/16cached.t            (Wstat: 256 (exited 1) Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
t/rt85886.t             (Wstat: 256 (exited 1) Tests: 2 Failed: 1)
  Failed test:  2
  Non-zero exit status: 1
Files=44, Tests=1488, 19 wallclock secs ( 0.25 usr  0.10 sys +  4.15 cusr  1.68 csys =  6.18 CPU)
Result: FAIL
Failed 2/44 test programs. 2/1488 subtests failed.
make: *** [Makefile:1114: test_dynamic] Error 1

ancientwizard committed:
Integrated SEGV testing into framework (namely lib/DBDOracleTestLib.pm)
  to make tests consistent in the opertion.
Spruced up tests to eliminate noise from the "ok's"

ancientwizard committed:
These calls randomly SEGV during logging in cnx_detach().
   OCISessionEnd_log_stat()
   OCIServerDetach_log_stat()

ancientwizard committed:
Error Detecton "BARKING" during threads / forking and OciSessionEnd()
  a balance between detecton, reporting and making the issue worse.

ancientwizard committed:
Consistent permissions, typos, test cleanups (EZCONNECT VS TRADITIONAL)

djzort committed:
Update Changes with recently merged pull requests

djzort committed:
Skip threads testing if threads arent compiled in to perl

djzort committed:
Try again to only run tests with threads are compiled in to perl

djzort committed:
Update Changes

djzort committed:
It is now 2025

djzort committed:
Bump to version 1.91_1

djzort committed:
Update MANIFEST file

ancientwizard committed:
Test cleanups.

I have fixed all DBD::Oracle bugs (I could detect). I have this remaining
  SEGV fault within Perl itself.

ADDED: Additional forking test that reproduces SEGV
  within Perl 5.30.3+ using Oracle 19, 21 & 23. I've only tested
  with these combinations.

$ gdb $(which perl)
 -- snip --
(gdb) run t/92-segv-fork.t
 -- snip --

Thread 2 "perl" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffff1cc4680 (LWP 9807)]
0x000055555562e610 in Perl_csighandler3 (sig=17, sip=0x0, uap=0x0) at mg.c:1601
1601	           (PL_signals & PERL_SIGNALS_UNSAFE_FLAG))
(gdb) list
1596	           sig == SIGSEGV ||
1597	#endif
1598	#ifdef SIGFPE
1599	           sig == SIGFPE ||
1600	#endif
1601	           (PL_signals & PERL_SIGNALS_UNSAFE_FLAG))
1602	        /* Call the perl level handler now--
1603	         * with risk we may be in malloc() or being destructed etc. */
1604	    {
1605	        if (PL_sighandlerp == Perl_sighandler)
(gdb) bt
    futex_word=0x5555563af118) at ./nptl/futex-internal.c:57
    at ./nptl/futex-internal.c:87
    abstime=abstime@entry=0x7ffff1cc3e00, private=private@entry=0) at ./nptl/futex-internal.c:139
    at ./nptl/pthread_cond_wait.c:503
(gdb)

ancientwizard committed:
Update 92-segv-fork.pl

The comments were typo "CITY"

djzort committed:
Update Changes file

djzort committed:
Bump to v1.19_2

djzort committed:
Update MANIFEST file

ancientwizard committed:
Just in case https://github.com/perl5-dbi/DBD-Oracle/issues/59 is right
  and someting is marginal. I didn't see the warnings and unable
  to locate a test that triggerd the use of this code path.

ancientwizard committed:
Added proof that connecting to the DB in the parent results in at
 least one additional non-Perl thread is created. If this thread
 is tapped to handle the SIGCHLD SEGV will result.

ancientwizard committed:
Update 92-segv-fork.t

Allow test not to fail on mistake

ancientwizard committed:
Re-applied test update to support OS's that dont support using
  USR1 & USR2; replaced with HUP and INT for OS's (Windows)

Confirmed that the Perl Interrupt patch and it's alternative work
suitably well at protecting against the SEGV caused by Oracle
instant client non-Perl worker threads.

ancientwizard committed:
I forgot the reason I had to disable this. The Perl being build in the GitHub test
does not have thread support built in.

lewart03 committed:
Use /sbin/ldconfig -p to find ORACLE_HOME

Incorporating @tux's feedback

djzort committed:
Avoid SvGROW underflow when phs->maxlen is zero GH105

Guard the call to `SvGROW` in `dbdimp.c` so we only grow the Perl SV
when `phs->maxlen > 0`. Previously `SvGROW(phs->sv, phs->maxlen-1)`
could underflow when `phs->maxlen` was 0 (e.g. bind_param_inout with
undef), producing a huge unsigned size and triggering Perl's "memory
wrap" panic (seen in GH#105 / RT94232). This adds a simple defensive
check and keeps behavior identical when `maxlen` is positive.

- File changed: `dbdimp.c`
- Key change: only call `SvGROW(phs->sv, ...)` for `phs->ftype == 96`
  when `phs->maxlen > 0`.

This fixes the panic observed when rebinding zero-length in/out
parameters and prevents passing an enormous value to `SvGROW`.

alexm committed:
Switch to gvenzl/oracle-free (Closes: #200)

djzort committed:
Update changes ahead of v1.91_3

djzort committed:
Bump to v1.91_3

djzort committed:
Update .mailmap

djzort committed:
Update changes

djzort committed:
Restore placeholder

djzort committed:
Fix some typos in comments

djzort committed:
Fix documentation copied from DBD::Pg (GH #115)

The bind_param examples and ParamTypes documentation contained
text copied from DBD::Pg that was never adapted for DBD::Oracle:

 - Replace pg_type attribute with ora_type in bind_param examples
 - Replace ORA_NUMBER (not a valid OCI bind type) with SQLT_INT
 - Replace PostgreSQL-style $1/$2 placeholders with positional integers
 - Pass SQL_INTEGER directly instead of wrapping in { ora_type => ... } since it is a DBI-level type constant, not an Oracle type
 - Fix ParamTypes docs referencing pg_type and "Postgres type"

Fixes: https://github.com/perl5-dbi/DBD-Oracle/issues/115

djzort committed:
Update Changes

djzort committed:
Document NOT INSTANTIABLE object type limitation (GH #16)

Selecting from a table whose column uses a NOT INSTANTIABLE parent type
fails with OCI-21500 when the result set contains rows of different
concrete subtypes interleaved with NULLs. Oracle Support confirmed this
requires using OCIObjectNew with the correct per-row TDO rather than
the parent type's TDO in OCIDefineObject, which would require
significant rearchitecting of the object fetch path.

Added a Known Limitations note to the ora_objects documentation and a
TODO test in 58object.t that reproduces the exact failure scenario.
The TODO test will automatically flag as an unexpected success if the
underlying OCI issue is fixed in the future.

Refs: https://github.com/perl5-dbi/DBD-Oracle/issues/16

djzort committed:
Replace old k&r style definitions to quiet down compiler warnings

djzort committed:
Replace GIMME with GIMME_V

From https://perldoc.perl.org/perl5380delta

```
The underlying Perl_dowantarray function implementing the long-deprecated GIMME macro has been marked as deprecated, so that use of the macro emits a compile-time warning. GIMME has been documented as deprecated in favour of GIMME_V since Perl v5.6.0, but had not previously issued a warning.
```
Anything that old we just don't care about. The Makefile.PL requires 5.6.0

djzort committed:
Update changes

lzsiga committed:
Change every occurance of 'sv_setpvn' to 'sv_setpvn_mg': the former might cause problem when utf-8 is used (pr-202)

djzort committed:
Refactor Oracle test script a tiny bit

djzort committed:
Adjust diag() calls to note() where appropriate

djzort committed:
maint/ Don't assume awk isn't there.

djzort committed:
maint/ upgrade the perl image base

djzort committed:
Add unit test for gh#86

djzort committed:
Skip 92-segv-fork.t unless on linux

djzort committed:
Update Changes

djzort committed:
Bump version to 1.91_4

djzort committed:
Update MANIFEST

djzort committed:
Refactor sample oracle xe install script

djzort committed:
Add README.md file in maint/

djzort committed:
Ignore META-INF garbage

djzort committed:
Adjust CMD format to clear docker build warnings

djzort committed:
Increase Oracle client version to 21.21

djzort committed:
Add version 21.21 and 23.26

djzort committed:
Libaio work around for trixie

From https://forums.oracle.com/ords/apexds/post/instant-client-on-ubuntu-24-04-noble-numbat-7244

djzort committed:
Increate to trixie and 23.26

djzort committed:
Fix stale SvCUR causing wrong output length for inout params gh#216

On Perl < 5.19.1 (before copy-on-write became default), the test
t/25plsql.t line 232 ('expected return length') fails because
length($p1) returns 5 instead of 200.

Root cause: when an SV is undef'd after holding a value (e.g. 'null!'
with SvCUR=5), SvCUR retains the stale length. SvGROW does not reset
SvCUR. When the inout buffer is grown to exactly match maxlen (200),
SvLEN also becomes 200. After Oracle writes 200 bytes (phs->alen=200),
the UNTOUCHED heuristic in dbd_phs_sv_complete() sees phs->alen ==
SvLEN(sv) and falsely concludes Oracle didn't touch the buffer, falling
back to SvCUR(sv) which is the stale value 5.

Perl 5.19.1+ masked this because COW changes how SvGROW allocates
buffers, making SvLEN differ from the requested size and avoiding
the false positive in the heuristic.

Two fixes applied:

1. dbd_rebind_ph_char(): After SvGROW for inout params, reset SvCUR
   to 0 when the SV is undef. This eliminates stale SvCUR values.

2. dbd_phs_sv_complete(): Add SvPOK(sv) guard to the UNTOUCHED
   heuristic. When an SV was undef (SvPOK off) before execute and
   Oracle reports indp==0, Oracle definitively wrote data, so trust
   phs->alen. The UNTOUCHED fallback now only applies when the SV
   had valid string content before execute.

Fixes: https://github.com/perl5-dbi/DBD-Oracle/issues/216

djzort committed:
Fix SIGCHLD race causing t/92-segv-fork.t to freeze (gh#212)

_SIG_CHLD called waitpid(-1, WNOHANG) exactly once per signal
delivery.  POSIX standard signals coalesce: if two or more children
exit between Perl safe-signal opchecks, only a single SIGCHLD is
queued.  The handler would then reap one child while the others
remained as zombies in $WORKSET, so running() never reached 0 and
the while(isBusy) loop in FORK_SEGV span forever.

Three changes:

1. Loop waitpid in _SIG_CHLD until no more children are available,
   draining all zombies per signal delivery.

2. Add a defensive reap() method called from the main polling loop,
   so exited children are collected even if a signal was lost or
   coalesced between opchecks.

3. Add a 300-second deadline to the main loop; call abort() instead
   of hanging indefinitely.

The same underlying bug (signal coalescing + single waitpid) was
also the likely cause of the Windows/Strawberry Perl hangs reported
in gh#191, although that issue has additional problems with missing
USR1/USR2 signals on Windows.

Fixes: gh#212
See-also: gh#191

djzort committed:
Fix SV refcount leak in pp_exec_rset() for unopened cursors

When a PL/SQL block returned a SYS_REFCURSOR that was never opened
(OCI_STMT_STATE_INITIALIZED), pp_exec_rset() replaced phs->sv with
newSV(0) without first decrementing the refcount on the original SV.
This leaked the user's bound scalar on every execute and silently
disconnected it from the placeholder, so subsequent re-executes of the
same statement wrote into an orphaned SV instead of the user's
variable.

Replace the `phs->sv = newSV(0)` assignment with `SvOK_off(phs->sv)`,
which clears the SV to undef in place.  This preserves the original
binding, avoids the refcount leak, and keeps re-execute working
correctly.

The OCI_STMT_STATE_INITIALIZED path was introduced for RT#82663 (null
cursor handling) but the leak has been present since that change.

Fixes: gh#217

djzort committed:
Support Oracle 23ai BOOLEAN type (gh#181)

Oracle Database 23ai introduces a native SQL BOOLEAN type with internal
OCI type code 252 (SQLT_BOL). Previously this type fell through to the
default handler, producing a spurious warning and reporting the wrong
DBI type code (-9252) in `$sth->{TYPE}`.

Changes:

- `dbdimp.c`: Add `#ifndef SQL_BOOLEAN` fallback (value 16) for older
  DBI installations. Add case 252 to `oratype_bind_ok()` so BOOLEAN
  columns are accepted as bind targets. Add `SQL_BOOLEAN -> 252`
  mapping in `ora_sql_type()` for DBI-standard bind type round-trip.
  Add `252 -> SQL_BOOLEAN` mapping in `ora2sql_type()` so
  `$sth->{TYPE}` correctly reports `SQL_BOOLEAN` (16).

- `oci8.c`: Add case 252 in `dbd_describe()` with `disize = 6` to
  accommodate the longest string representation ("FALSE\0") when OCI
  converts SQLT_BOL to a null-terminated string, avoiding a buffer
  overflow.

- `Oracle.xs`: Export `ORA_BOOLEAN = 252` as a named constant.

- `lib/DBD/Oracle.pm`: Add `ORA_BOOLEAN` to the `:ora_types` export
  tag and update all documentation sections listing Oracle type
  constants. Add a description of the type and its SQL_BOOLEAN mapping.

- `t/20select.t`: Add 5 tests in a version-gated SKIP block
  (requires Oracle >= 23) covering the ORA_BOOLEAN constant value,
  `$sth->{TYPE}` reporting SQL_BOOLEAN, and fetched true/false values.

djzort committed:
Only run t/92-segv-fork.t on perl 5.42 or later

See gh#212 for full details. But there is a bug in perl prior to this.

djzort committed:
Harden segv tests: Linux-only, version guards, bug fixes (gh#215)

## Overview

Address test hangs reported in gh#215 (and related gh#191, gh#212) where
`t/90-segv-threads.t` and `t/91-segv-fork.t` hang or crash on Windows
and older Perls. All three segv tests now run only on Linux, with
additional guards for minimum dependency versions.

## Changes

- **t/90-segv-threads.t**
  - Add Linux-only skip guard (`$^O ne 'linux'`)
  - Require `Thread::Queue 3.07` to avoid objects arriving as unblessed references across thread boundaries on older versions
  - Add 30-second timeout to thread join loop in `disable()` to prevent indefinite hangs when threads terminate abnormally

- **t/91-segv-fork.t**
  - Add Linux-only skip guard (`$^O ne 'linux'`)
  - Fix bug in `disable()` where `kill('USR2', ...)` was hardcoded instead of using the `$SIG_EXIT` variable, ignoring the signal fallback logic

- **t/92-segv-fork.t**
  - Add Linux-only skip guard (prior change)
  - Add Perl 5.42+ version requirement skip guard (prior change)

djzort committed:
Expand ora_events documentation with usage examples (gh#176)

- Add code examples for enabling FAN via connect attribute and
  environment variable
- Explain what FAN does (HA events from Oracle RAC clusters)
- Show common usage pattern combining `ora_events` with `ora_drcp`
- Clarify that `ora_events` is **not** related to Oracle Database
  Change Notification (DCN/CQN)
- Remove duplicate `ORA_EVENTS` environment variable check

djzort committed:
Remove unreachable code

djzort committed:
Fix join inside loop in table_info type filtering

The join() call was inside the foreach loop, wastefully re-joining
on every iteration. It produced correct results only by accident
on the final iteration. Move it after the loop so all elements are
quoted before joining.

Also simplify the quoting expression using string interpolation.

djzort committed:
Fix placeholder mismatch in dbms_msgpipe_ack

The SQL uses :errormsg but the bind used :proc, a copy-paste error
from dbms_msgpipe_get. This caused dbms_msgpipe_ack to always fail
at runtime since :proc doesn't exist in the acknowledge statement.

Remove debug line

djzort committed:
Complete rt91217 fix: remove deprecated CHOOSE hint from all queries

The original rt91217 fix (v1.71_00) only worked in statistics_info.
In column_info, an elsif chained to if (>= 8) made the >= 11 branch
unreachable, so the hint was never removed. In table_info,
primary_key_info, and foreign_key_info the hint was hardcoded in
single-quoted heredocs and never addressed at all.

- column_info: change elsif to standalone if so hint is cleared for 11+
- table_info, primary_key_info, foreign_key_info: remove hardcoded hint

djzort committed:
Add TLS/SSL (TCPS) connection documentation (#166)

Document how to connect to Oracle over TLS/SSL using TCPS protocol.
TLS is handled by the Oracle client libraries (OCI) and works
transparently once the client-side sqlnet.ora and wallet/certificates
are configured.

Covers Oracle Wallet setup, sqlnet.ora configuration, system
certificates (19c+), TNS entries with TCPS, and inline connect
descriptors.

djzort committed:
Fix ora_server_version() returning undef on first call (gh#205)

ora_server_version() computed the version from v$version and cached
it in $dbh->{ora_server_version}, but fell off the end of the sub
without returning the value.  Only the second call (which hit the
cache) returned correctly.

All four internal callers dereference the result as an arrayref
(e.g. ora_server_version($dbh)->[0]), so the first call on any
connection where the XS layer had not already populated the cache
would die with "Can't use an undefined value as an ARRAY reference".

Add the missing return statement, replacing the TODO comment that
already identified the bug.

Fixes: gh#205

djzort committed:
Correct skip_all behaviour when db connection not present

alexm committed:
Create ORACLE_USERID_2 for t/60reauth.t (Fixes: #211)

djzort committed:
Fix ORA_DATE bind param support (fixes GH#84)

Binding a placeholder with `ora_type => ORA_DATE` previously caused a
fatal error:

> Can't bind :p1, ora_type 12 not supported by DBD::Oracle

`ORA_DATE` (type 12) was missing from the `oratype_bind_ok()` whitelist
in `dbdimp.c`. The bound value is handled as a string by
`dbd_rebind_ph_char()` (the default path), which correctly passes it to
Oracle for implicit NLS date conversion — consistent with how the fetch
side already handles `ORA_DATE`.

- **dbdimp.c** — Added `case 12: /* DATE */` to `oratype_bind_ok()` switch
- **t/95-gh84_ora_date_bind.t** — New test: verifies `bind_param` with `ora_type => ORA_DATE` no longer croaks, executes an INSERT, and round-trips the value via SELECT
- **MANIFEST** — Added `t/95-gh84_ora_date_bind.t` (and two previously missing `t/94-*` entries)

djzort committed:
Bump to v1.91_5

djzort committed:
Bump minimum DBD to 1.632 circa 2014

djzort committed:
More nuanced DBI suggestions and requirements as recommended in GH#215

djzort committed:
Bump to v1.95

alexm committed:
Wrap skip_all in a BEGIN block (Closes: #219)

Thus making sure that the test is skipped before touching signals on
unsupported platforms.

alexm committed:
Remove $syslibs tweaking in Debian

It was needed with older versions of Oracle client to hardcode its
library path into the binary, but newer Oracle client versions
install `/etc/ld.so.conf.d/oracle-instantclient.conf` with the
library path, so there's no need to add anything in `$syslibs`
for Debian anymore.

Closes: #221
